src/to_markdown.js

```diff
@@ -219,75 +233,59 @@ export class MarkdownSerializerState {
-          if (!node) marks = active
-        }
-      }
-
-      let inner = marks.length && marks[marks.length - 1], noEsc = inner && this.marks[inner.type.name].escape === false
-      let len = marks.length - (noEsc ? 1 : 0)
-
       // Try to reorder 'mixable' marks, such as em and strong, which
       // in Markdown may be opened and closed in different order, so
       // that order of the marks for the token matches the order in
       // active.
-      outer: for (let i = 0; i < len; i++) {
+      outer: for (let i = 0; i < marks.length; i++) {
         let mark = marks[i]
         if (!this.marks[mark.type.name].mixable) break
         for (let j = 0; j < active.length; j++) {
-          let other = active[j]
+          let other = active[j].mark
           if (!this.marks[other.type.name].mixable) break
           if (mark.eq(other)) {
             if (i > j)
-              marks = marks.slice(0, j).concat(mark).concat(marks.slice(j, i)).concat(marks.slice(i + 1, len))
+              marks = marks.slice(0, j).concat(mark).concat(marks.slice(j, i)).concat(marks.slice(i + 1, marks.length))
             else if (j > i)
-              marks = marks.slice(0, i).concat(marks.slice(i + 1, j)).concat(mark).concat(marks.slice(j, len))
+              marks = marks.slice(0, i).concat(marks.slice(i + 1, j)).concat(mark).concat(marks.slice(j, marks.length))
             continue outer
           }
         }
       }
 
       // Find the prefix of the mark set that didn't change
       let keep = 0
-      while (keep < Math.min(active.length, len) && marks[keep].eq(active[keep])) ++keep
+      while (keep < Math.min(active.length, marks.length) && marks[keep].eq(active[keep].mark)) ++keep
 
-      // Close the marks that need to be closed
-      while (keep < active.length)
-        this.text(this.markString(active.pop(), false), false)
+      // For each mark that needs to be closed, write its delimiters and accumulated content into its parent state
+      while (keep < active.length) {
+        const {mark, parentState} = active.pop()
 
-      // Output any previously expelled trailing whitespace outside the marks
-      if (leading) this.text(leading)
+        parentState.text(parentState.markString(mark, currentState.out), false)
+        currentState = parentState
+      }
 
-      // Open the marks that need to be opened
+      // For each mark that needs to be opened, create a new state to accumulate its content
+      // and keep track of its parent state to write the delimiters and content into later
       if (node) {
-        while (active.length < len) {
-          let add = marks[active.length]
-          active.push(add)
-          this.text(this.markString(add, true), false)
+        while (active.length < marks.length) {
+          active.push({mark: marks[active.length], parentState: currentState})
+
+          currentState = new MarkdownSerializerState(this.nodes, this.marks, this.options)
```

Comment:
Since the code mark opening and closing delimiters depend on their serialized content of the mark, we cannot write anything onto `this.state` until the mark is actually ready to be closed, we know what its serialized content looks like, and we can determine the appropriate delimiters based on it. 

`currentState` functions as a scratch pad of sorts, to keep track of the serialized content of the mark. `currentState.out` is passed to `markString` as `text` and written to `this.state` along with the dynamically determined opening and closing delimiters.

Nothing is "different about the copies" beyond the fact that their `out` starts out empty, and that they are indeed different copies :)

Do you think it would help if it was called something like `markInnerState` or `markContentState`?

Reply:
Why does the content have to be written out in advance? Since inline code doesn't support any kind of escaping, isn't the plain content enough to base the delimiters on?

Reply:
I'd prefer to keep things as simple as reasonable, so I'd rather drop the multiple state approach and just make things work for code spans.

Reply:
@marijnh All right. 

If I simplify this PR to focus solely on the code span case, would you be open to discussing the multiple-state approach some more in a new PR that will focus specifically on serializing autolinks as `<url>` rather than `[url](url)`? I'm looking forward to hearing other approaches you may have in mind to realize this behavior. Ultimately, my goal is to get prosemirror-markdown serialized Markdown to be as close to human-written Markdown as possible :)

Reply:
I'm not sure you need multiple states for that case either—you can define link formatting on top of the link's content (as ProseMirror document nodes), you don't need its Markdown text there, do you?

Reply:
@marijnh I'm not sure I follow; are you suggesting using an inline node for links, instead of a mark?

Reply:
No, I'm suggesting giving the function that produces the delimiters access to the inline node(s) that the mark applies to.

Reply:
@marijnh All right, I'll simplify this PR for the code mark case, and I'll give that a try / we can discuss that some more in the follow-up PR I'll create for autolinks!